### PR TITLE
Fix mixed open and close generic service IEnumerable resolution

### PR DESCRIPTION
**Pull request by pakrym**
Fixes: #500

### CODE REVIEW

**Comment by pakrym** [comment]
Before:

```
    Method |        Mean |    StdErr |     StdDev |      Median |         Op/s | Scaled | Scaled-StdDev |  Gen 0 | Allocated |
---------- |------------ |---------- |----------- |------------ |------------- |------- |-------------- |------- |---------- |
      NoDI |   5.5142 ns | 0.0348 ns |  0.1908 ns |   5.4640 ns | 181348688.78 |   1.00 |          0.00 | 0.0035 | 24 B |
 Transient |  55.4727 ns | 0.3655 ns |  2.0019 ns |  54.8612 ns |  18026897.93 |  10.07 |          0.48 | 0.0015 | 24 B |
    Scoped | 118.1930 ns | 0.7217 ns |  3.9530 ns | 117.7722 ns |   8460736.43 |  21.46 |          0.99 |      - | 0 B |
 Singleton | 157.7186 ns | 5.1082 ns | 27.9789 ns | 174.8609 ns |   6340408.26 |  28.63 |          5.08 |      - | 0 B |
```
After:
```
    Method |        Mean |    StdDev |         Op/s | Scaled | Scaled-StdDev |  Gen 0 | Allocated |
---------- |------------ |---------- |------------- |------- |-------------- |------- |---------- |
      NoDI |   5.9400 ns | 0.1832 ns | 168350433.44 |   1.00 |          0.00 | 0.0035 |      24 B |
 Transient |  56.2400 ns | 1.4699 ns |  17780950.96 |   9.48 |          0.37 | 0.0015 |      24 B |
    Scoped | 128.1737 ns | 3.9677 ns |   7801911.84 |  21.60 |          0.91 |      - |       0 B |
 Singleton | 135.9294 ns | 3.8256 ns |   7356759.37 |  22.90 |          0.92 |      - |       0 B |
```

File: src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteFactory.cs [hunk 1]
```diff
@@ -0,0 +1,346 @@
...
+                var closedType = descriptor.ImplementationType.MakeGenericType(type.GenericTypeArguments);
+                var constructorCallSite = CreateConstructorCallSite(type, closedType, callSiteChain);
+
+                return ApplyScope(constructorCallSite, descriptor.Lifetime);
+            }
+
+            return null;
+        }
+
+        private IServiceCallSite ApplyScope(IServiceCallSite serviceCallSite, ServiceLifetime descriptorLifetime)
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteFactory.cs]: Change the name to ApplyLifetime, Scope has another meaning

File: src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteFactory.cs [hunk 2]
```diff
@@ -0,0 +1,346 @@
...
+        private readonly List<ServiceDescriptor> _descriptors;
+        private readonly Dictionary<Type, IServiceCallSite> _callSiteCache = new Dictionary<Type, IServiceCallSite>();
+
+        public CallSiteFactory(IEnumerable<ServiceDescriptor> descriptors)
+        {
+            _descriptors = descriptors.ToList();
+            Validate(descriptors);
+        }
+
+        private void Validate(IEnumerable<ServiceDescriptor> descriptors)
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyInjection/ServiceLookup/CallSiteFactory.cs]: static?

File: src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs [hunk 3]
```diff
@@ -680,5 +680,29 @@ public void DisposesInReverseOrderOfCreation()
             Assert.Equal(outer.MultipleServices.Reverse(), callback.Disposed.Skip(1).Take(3).OfType<IFakeMultipleService>());
             Assert.Equal(outer.SingleService, callback.Disposed[4]);
         }
+
+        [Fact]
+        public void ResolvesMixedOpenClosedGenericsAsEnumerable()
```
Comment (pakrym) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: /cc @khellang this test fails on StructureMap

File: src/Microsoft.Extensions.DependencyInjection/ServiceLookup/FactoryService.cs [hunk 4]
```diff
@@ -2,33 +2,20 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
 using System;
-using System.Collections.Generic;
 
 namespace Microsoft.Extensions.DependencyInjection.ServiceLookup
 {
-    internal class FactoryService : IService, IServiceCallSite
+    internal class FactoryService : IServiceCallSite
```
Comment (davidfowl) [inline src/Microsoft.Extensions.DependencyInjection/ServiceLookup/FactoryService.cs]: FactoryCallsite

File: src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs (hunk 3 above)
Comment (khellang) [inline src/Microsoft.Extensions.DependencyInjection.Specification.Tests/DependencyInjectionSpecificationTests.cs]: Yep. That was expected (https://github.com/aspnet/DependencyInjection/issues/500#issuecomment-288051146). I'll see if I can do some magic internally to get it to work.

**Review by davidfowl — APPROVED** [review]

**Pull request merged**